src/EffectiveCSharp.Analyzers/PreferDeclarationInitializersToAssignmentStatementsAnalyzer.cs

```diff
@@ -0,0 +1,494 @@
+        isEnabledByDefault: true,
+        description: "Field initialization to null or zero is redundant and should be avoided.",
+        helpLinkUri: HelpLinkUri);
+
+    private static readonly DiagnosticDescriptor RuleExceptionShouldNotInitializeInDeclaration = new(
+        id: DiagnosticIds.PreferDeclarationInitializersExceptWhenVaryingInitializations,
+        title: "Should not initialize in declaration due to diverging initializations in constructors",
+        messageFormat: "Do not initialize a field in its declaration if you have diverging initializations in constructors. This is to prevent unnecessary allocations.",
+        category: "Maintainability",
+        defaultSeverity: DiagnosticSeverity.Warning,
+        isEnabledByDefault: true,
+        description: "Field initialization should not occur when there are diverging initializations in constructos. This is to prevent unnecessary allocations.",
+        helpLinkUri: HelpLinkUri);
+
+    private static readonly DiagnosticDescriptor RuleShouldInitializeInDeclarationWhenNoInitializationPresent = new(
+        id: DiagnosticIds.PreferDeclarationInitializersWhenNoInitializationPresent,
+        title: "Should initialize in declaration when no initialization present",
+        messageFormat: "Initialize the field in its declaration when no distint initializations will occur in constructors",
+        category: "Maintainability",
+        defaultSeverity: DiagnosticSeverity.Warning,
+        isEnabledByDefault: true,
+        description: "Field initialization should occur in the declaration unless there are diverging initializations in constructors, or the field is a value type or nullable being initialized to the default.",
+        helpLinkUri: HelpLinkUri);
+
+    /// <inheritdoc />
+    public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(GeneralRule, RuleExceptionShouldNotInitializeToNullOrZero, RuleExceptionShouldNotInitializeInDeclaration, RuleShouldInitializeInDeclarationWhenNoInitializationPresent);
+
+    /// <inheritdoc />
+    public override void Initialize(AnalysisContext context)
+    {
+        context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+        context.EnableConcurrentExecution();
+        context.RegisterSyntaxNodeAction(AnalyzeNode, SyntaxKind.ClassDeclaration);
+    }
+
+    private static void AnalyzeNode(SyntaxNodeAnalysisContext context)
+    {
+        ClassDeclarationSyntax classSyntaxNode = (ClassDeclarationSyntax)context.Node;
+
+        // In order to keep track of the fields that are initialized in the constructors
+        // We create this dictionary that will track a field's initialization in the constructors.
+        // To save on time and memory, we only track fields that are initialized in the constructors.
+        IDictionary<string, FieldInitializationInfo> fieldInitializationInfo = new Dictionary<string, FieldInitializationInfo>(StringComparer.Ordinal);
```

Comment:
No need to use interface here, it doesn't buy you anything. Use the concrete type `Dictionary`